hedera-node/src/main/java/com/hedera/services/txns/contract/ContractUpdateTransitionLogic.java

```diff
@@ -21,77 +21,122 @@
 	) {
-		this.ledger = ledger;
-		this.validator = validator;
 		this.txnCtx = txnCtx;
-		this.contracts = contracts;
-		this.customizerFactory = customizerFactory;
+		this.validator = validator;
+		this.accountStore = accountStore;
 	}
 
 	@Override
 	public void doStateTransition() {
-		try {
-			var contractUpdateTxn = txnCtx.accessor().getTxn();
-			var op = contractUpdateTxn.getContractUpdateInstance();
-			var id = op.getContractID();
-			var target = contracts.get().get(fromContractId(id));
-
-			var result = customizerFactory.customizerFor(target, validator, op);
-			var customizer = result.getLeft();
-			if (customizer.isPresent()) {
-				ledger.customize(asAccount(id), customizer.get());
-				txnCtx.setStatus(SUCCESS);
-			} else {
-				txnCtx.setStatus(result.getRight());
-			}
-		} catch (Exception e) {
-			log.warn("Avoidable exception!", e);
-			txnCtx.setStatus(FAIL_INVALID);
+		/* --- Extract gRPC --- */
+		final var contractUpdateTxn = txnCtx.accessor().getTxn();
+		final var op = contractUpdateTxn.getContractUpdateInstance();
+		final var targetId = Id.fromGrpcContract(op.getContractID());
+
+		/* --- Translate the account --- */
+		final var target = accountStore.loadContract(targetId);
+		final var candidate = op.getAdminKey();
+
+		/* --- Conditions --- */
+		final var isSentinel = candidate.equals(IMMUTABILITY_SENTINEL_KEY);
+
+		/* --- Validate --- */
+		if (op.hasAdminKey()) {
+			validateFalse(isValidAdminKey(candidate, isSentinel), INVALID_ADMIN_KEY);
+		}
+		if (op.hasExpirationTime()) {
+			validateFalse(!validator.isValidExpiry(op.getExpirationTime()), INVALID_EXPIRATION_TIME);
+		}
+
+		/* --- Extract --- */
+		final var newAdminKey = getKey(candidate, isSentinel, target);
+		final var newProxy = op.hasProxyAccountID() ? op.getProxyAccountID() : null;
+		final var newAutoRenewPeriod = op.hasAutoRenewPeriod() ? op.getAutoRenewPeriod() : null;
+		final var newExpirationTime = op.hasExpirationTime() ? op.getExpirationTime() : null;
+		final var newMemo = affectsMemo(op) ? getMemo(op) : null;
+
+		/* --- Do the business logic --- */
+		target.updateFromGrpcContract(
+				Optional.ofNullable(newAdminKey),
+				Optional.ofNullable(newProxy),
+				Optional.ofNullable(newAutoRenewPeriod),
+				Optional.ofNullable(newExpirationTime),
+				Optional.ofNullable(newMemo),
+				affectsExpiryOnly(op));
+
+		/* --- Persist the changes --- */
+		accountStore.persistUpdatedContract(target);
+	}
+
+	private boolean isValidAdminKey(final Key candidate, final boolean isSentinelKey) {
+		if (isSentinelKey) {
+			return false;
+		} else {
+			var key = MiscUtils.asUsableFcKey(candidate);
+			return key.isEmpty() || key.get() instanceof JContractIDKey;
 		}
 	}
 
+	private JKey getKey(final Key candidate, final boolean isSentinelKey, final Account target) {
+		return isSentinelKey
+				? new JContractIDKey(target.getId().getShard(), target.getId().getRealm(), target.getId().getNum())
+				: MiscUtils.asUsableFcKey(candidate).orElse(null);
```

Comment:
Why don't you simply return an Optional.of(whatever-comes-in-here-after-refactor) and pass it directly into the updateContract method as it needs Optional parameters anyway?